### ISSUE TITLE
twisted.python.urlpath.URLPath doesn't respect the default path of '/'

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8068]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8068|
|Type|release blocker: regression|
|Created|2015-10-27 00:21:58Z|
|Branch|https://github.com/twisted/twisted/tree/urlpath-str-8068-3|

In 15.4:
```
>>> URLPath.fromString("http://google.com")
URLPath(scheme='http', netloc='google.com', path='/', query='', fragment='')
>>> str(URLPath.fromString("http://google.com"))
'http://google.com/'
```

...in trunk...

```
>>> URLPath.fromString("http://google.com")
URLPath(scheme='http', netloc='google.com', path='', query='', fragment='')
>>> str(URLPath.fromString("http://google.com"))
'http://google.com'
```

<details><summary>Searchable metadata</summary>

```
trac-id__8068 8068
type__release_blocker__regression release blocker: regression
reporter__hawkowl hawkowl
priority__highest highest
milestone__Python_3_x Python-3.x
branch__branches_urlpath_str_8068_3 branches/urlpath-str-8068-3
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1445905318738929 1445905318738929
changetime__1447413352461878 1447413352461878
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [45993]) Branching to urlpath-str-8068.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

(In [45999]) Branching to urlpath-str-8068-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

Builders spun, please review.

bee tee dubs, this makes the Treq testsuite pass on Python 3.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @hawkowl|
|-|-|

From https://github.com/twisted/twisted/compare/trunk...urlpath-str-8068-2 -

"0 files changed"

probably need to fix that.

**Comment by twisted-trac** [comment]
|<img alt="posita's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|posita commented|
|-|-|

I'm not sure where to present this comment. My apologies in advance is there is a more appropriate place.

The `URLPath.fromBytes` implementation does not play well with [http://python-future.org/reference.html?highlight=newbytes#future.types.newbytes future's `newbytes`] type:

```
Python 2.7.10 (default, Sep 24 2015, 10:13:45)
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from __future__ import unicode_literals
>>> from twisted.python.urlpath import URLPath
>>> from future.types import newbytes, newstr
>>> u1 = URLPath.fromString(u'http://foo.com/')
>>> type(u1.scheme)
<type 'str'>
>>> u2 = URLPath.fromString(newstr(u'http://foo.com/'))
>>> type(u2.scheme)
<type 'str'>
>>> u3 = URLPath.fromBytes(b'http://foo.com/')
>>> type(u3.scheme)
<type 'str'>
>>> u4 = URLPath.fromBytes(newbytes(b'http://foo.com/'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../lib/python2.7/site-packages/twisted/python/urlpath.py", line 131, in fromBytes
    return klass(*parts)
  File "/.../lib/python2.7/site-packages/twisted/python/urlpath.py", line 50, in __init__
    self._url = _URL.fromText(urltext.encode("ascii").decode("ascii"))
  File "/.../lib/python2.7/site-packages/future/types/newbytes.py", line 366, in __getattribute__
    raise AttributeError("encode method has been disabled in newbytes")
AttributeError: encode method has been disabled in newbytes
```

This is primarily because of [https://github.com/twisted/twisted/blob/31e18b0472ffe02b221e3cda1614f90fe181b1b8/twisted/python/urlpath.py#L50 this trick in the `URLPath` constructor]:

```
        # ...
        urltext = urlquote(
            urlparse.urlunsplit((self.scheme, self.netloc,
                                 self.path, self.query, self.fragment)),
            safe=_allascii
        )
        self._url = _URL.fromText(urltext.encode("ascii").decode("ascii"))
        # ...
```

This assumes (takes advantage of the fact) that `urltext` (i.e., the returned value from `urlquote(urlunplit(...))`) is normally a "native" string type (i.e., a byte string in Python 2, but a Unicode string in Python 3). The Python 2 native string type (perversely) has _both_ an `encode` method _and_ a `decode` method. This is not the case with either Python 3 types (Unicode strings have an `encode` method, byte strings have a `decode` method), nor is it the case with future's `newbytes`:

```
>>> from twisted.python.compat import urllib_parse as urlparse, urlquote
>>> from twisted.python.urlpath import _allascii
>>> u5 = urlparse.urlunsplit(( newbytes(b'http'), newbytes(b'foo.com'), newbytes(b'/'), newbytes(b''), newbytes(b'') ))
>>> u5 ; type(u5)
b'http://foo.com/'
<class 'future.types.newbytes.newbytes'>
>>> u6 = urlquote(u5, safe=_allascii)
>>> u6 ; type(u6)
b'http://foo.com/'
<class 'future.types.newbytes.newbytes'>
>>> u6.encode(u'ascii').decode(u'ascii')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../lib/python2.7/site-packages/future/types/newbytes.py", line 366, in __getattribute__
    raise AttributeError("encode method has been disabled in newbytes")
AttributeError: encode method has been disabled in newbytes
```

The only reason I bring this up is `future`'s apparent popularity in creating cross-compatible code (which is why I use it). I understand if supporting such a thing is beyond the scope of Twisted.

However, `<str>.encode('ascii').decode('ascii')` is a hack (and an artifact of a) Python 2's broken notion that you can encode a byte string or decode a unicode string; and b) trying to write one set of code that works for two distinct type of string types and two separate standard libraries). This can get you into trouble if you're not careful (see, e.g., [#8044](https://github.com/twisted/twisted/issues/8044)).

This can be worked around of course, but the work-around is not trivial (since `newbytes` works some ABC magic to appear as `bytes` in many cases):

```
>>> isinstance(b'', newbytes)
True
>>> isinstance(newbytes(b''), bytes)
True
```

One could do this:

```
class FutureCompatibleURLPath(URLPath):
    @classmethod
    def fromBytes(cls, url):
        try:
            native_url = url.__native__()
        except AttributeError:
            native_url = url
        return super(FutureCompatibleURLPath, cls).fromBytes(native_url)
```

Or simply:

```
class FutureCompatibleURLPath(URLPath):
    @classmethod
    def fromBytes(cls, url_bytes):
        return cls.fromString(url_bytes.decode('ascii'))

```

**Comment by twisted-trac** [comment]
|<img alt="posita's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|posita commented|
|-|-|

Actually, that does prompt me to ask, what about this as an alternate implementation?

```
class URLPath(object):
    # ...
    @classmethod
    def fromBytes(klass, url): # adopting existing naming conventions
        """
        ...
        """
        if not isinstance(url, bytes):
            raise ValueError("'url' must be bytes")
        return klass.fromString(url_bytes.decode('ascii'))
    # ...
```

It's possible with the above implementation that one would get a stack where `fromString` called `fromBytes`, which then called `fromString` again, but it _should_ end there. Of course, one would have to be a little careful to avoid infinite recursion when overriding one or both of `fromBytes` and `fromString`.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl removed owner|
|-|-|

Okay, it's actually had builders spun now, putting up for review.

RE: posita, we document that it takes a ```bytes```, not python future's newbytes type (which is an entirely different type that is incompatible). We use the "encoding a bytestring" trick in a few places to ensure that it only contains encodable things -- where otherwise this check would go undone.

**Comment by twisted-trac** [comment]
|<img alt="posita's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|posita commented|
|-|-|

I get that. My point is that `bytes` are not `encode`-able. The fact that one can technically call `encode` on a byte string in Python 2 doesn't mean Twisted should. `;o)`

Any thoughts on my suggestion in comment:6?

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Changes looks good, but I think the conversion needs a in-code comment.

I find it confusing to have bytes... and then process them with urlparse without specifying an encoding ... and get encoded str out of that operation

---------

The drive-by tests test_partialArguments might be ok... but  I find it confusing. To me it looks like 2 tests test_withoutArguments and test_partialArguments


please see my comment and merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl removed owner|
|-|-|

trunk moved in the meantime and the fix had to be rewritten. I wrote the tests, glyph provided the majority of the fix. Tests are looking ok. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban commented|
|-|-|

Is the ticket description and title still valid?

What was the problem and what is the expected result?

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban set owner to @hawkowl|
|-|-|

OK. All good

I tried on this branch on both py2 and py3 and it looks good.

```
>>> from twisted.python.urlpath import URLPath
>>> str(URLPath.fromString("http://google.com"))
'http://google.com/'
>>> 

```

please merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_15"></a>|@hawkowl set status to `closed`|
|-|-|

(In [46176]) Merge urlpath-str-8068-3: Make twisted.python.urlpath.URLPath respect the default path of '/'

Author: hawkowl
Reviewers: adiroiban
Fixes: [#8068](https://github.com/twisted/twisted/issues/8068)
